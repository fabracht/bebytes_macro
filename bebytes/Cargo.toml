[package]
name = "bebytes"
version = "0.7.0"
edition = "2021"
license = "MIT"
repository = "https://github.com/fabracht/bebytes_macro"
description = "A Rust library for serialization and deserialization of network structs."
readme = "README.md"
keywords = ["serialization", "deserialization", "network", "bytes", "no-std"]


[[bin]]
name = "macro_test"
path = "./bin/macro_test.rs"

[dependencies]
bebytes_derive = "0.8.0"

[dev-dependencies]
trybuild = { version = "1.0.102", features = ["diff"] }
test-case = "3.3.1"

[features]
default = ["std"]
std = []
