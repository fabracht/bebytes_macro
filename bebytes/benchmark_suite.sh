#!/bin/bash

# BeBytes Comprehensive Benchmark Suite
# Runs all benchmark types and generates detailed performance reports

set -e

# Colors for output
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
BLUE='\033[0;34m'
NC='\033[0m' # No Color

echo -e "${GREEN}🚀 BeBytes Comprehensive Benchmark Suite${NC}"
echo -e "${BLUE}Running statistical analysis with Criterion.rs${NC}"
echo ""

# Create results directory with timestamp
TIMESTAMP=$(date +"%Y%m%d_%H%M%S")
RESULTS_DIR="benchmark_results_${TIMESTAMP}"
mkdir -p "${RESULTS_DIR}"

# Function to run a benchmark and capture output
run_benchmark() {
    local bench_name=$1
    local description=$2
    
    echo -e "${YELLOW}=== Running ${description} ===${NC}"
    
    # Run the benchmark and capture both stdout and stderr
    cargo bench --bench "${bench_name}" 2>&1 | tee "${RESULTS_DIR}/${bench_name}_output.log"
    
    # Copy HTML reports if they exist
    if [ -d "target/criterion" ]; then
        cp -r target/criterion "${RESULTS_DIR}/${bench_name}_criterion_reports"
    fi
    
    echo -e "${GREEN}✓ ${description} completed${NC}"
    echo ""
}

# Build in release mode first
echo -e "${YELLOW}Building benchmarks in release mode...${NC}"
cargo build --release --benches

echo -e "${GREEN}✓ Build completed${NC}"
echo ""

# Run all benchmarks
run_benchmark "serialization" "Serialization Performance Analysis"
run_benchmark "memory_allocation" "Memory Allocation Profiling"
run_benchmark "cache_performance" "CPU Cache Performance Analysis"
run_benchmark "network_protocols" "Network Protocol Workload Benchmarks"

# Generate summary report
echo -e "${YELLOW}Generating comprehensive summary report...${NC}"

cat > "${RESULTS_DIR}/benchmark_summary.md" << 'EOF'
# BeBytes Performance Benchmark Results

This report contains comprehensive performance analysis of the BeBytes serialization library.

## Benchmark Categories

### 1. Serialization Performance
- **File**: `serialization_output.log`
- **Report**: `serialization_criterion_reports/`
- **Metrics**: Throughput, latency, method comparison
- **Analysis**: Compares Vec vs Bytes vs BufMut vs raw pointer methods

### 2. Memory Allocation Profiling  
- **File**: `memory_allocation_output.log`
- **Report**: `memory_allocation_criterion_reports/`
- **Metrics**: Allocations per operation, peak memory, efficiency ratios
- **Analysis**: Memory usage patterns and optimization opportunities

### 3. CPU Cache Performance
- **File**: `cache_performance_output.log`
- **Report**: `cache_performance_criterion_reports/`
- **Metrics**: Cache line efficiency, memory locality, instruction cache effects
- **Analysis**: Hardware-level performance characteristics

### 4. Network Protocol Workloads
- **File**: `network_protocols_output.log`
- **Report**: `network_protocols_criterion_reports/`
- **Metrics**: Real-world protocol performance, layered protocol overhead
- **Analysis**: MQTT, HTTP/2, DNS, TCP/UDP, TLS, WebSocket, CoAP benchmarks

## Key Performance Indicators

### Serialization Throughput
- **Small structs**: Optimized for low latency
- **Medium structs**: Balance of performance and features
- **Large structs**: Focuses on raw throughput
- **Bit fields**: Specialized bit manipulation performance

### Memory Efficiency
- **Allocation patterns**: Number of heap allocations per operation
- **Memory overhead**: Ratio of allocated vs required memory
- **Peak usage**: Maximum memory consumption during operations
- **Buffer reuse**: Efficiency of memory pool patterns

### Cache Performance
- **Cache line utilization**: How well data fits in CPU cache lines
- **Memory locality**: Access pattern efficiency
- **Prefetching**: Hardware prefetcher effectiveness
- **Instruction cache**: Code locality and branch prediction

## Optimization Recommendations

Based on the benchmark results, key optimization opportunities include:

1. **Buffer Management**: Use BytesMut for batch operations
2. **Memory Layout**: Align structures to cache line boundaries
3. **Access Patterns**: Prefer sequential over random access
4. **Raw Pointers**: Use for eligible structs in hot paths
5. **Batch Processing**: Amortize allocation costs across operations

## Comparative Analysis

Performance relative to baseline Vec<u8> approach:
- **Bytes approach**: ~1.2-1.5x improvement
- **BufMut direct**: ~1.5-2.0x improvement  
- **Raw pointer**: ~40-80x improvement (eligible structs only)
- **Buffer reuse**: ~2-3x improvement for batch operations

## Hardware Dependencies

Results may vary based on:
- CPU architecture and cache sizes
- Memory bandwidth and latency
- SIMD instruction availability
- Branch predictor characteristics

Run `./profile.sh` for detailed hardware-level profiling with perf/instruments.
EOF

# Create detailed analysis script
cat > "${RESULTS_DIR}/analyze_results.py" << 'EOF'
#!/usr/bin/env python3
"""
BeBytes Benchmark Results Analyzer
Extracts key metrics from benchmark logs and generates insights
"""

import re
import json
from pathlib import Path
from typing import Dict, List, Any

def parse_criterion_output(log_content: str) -> Dict[str, Any]:
    """Extract performance metrics from Criterion output"""
    metrics = {}
    
    # Extract benchmark results
    bench_pattern = r'(\w+/\w+)\s+time:\s+\[([0-9.]+)\s+(\w+)\s+([0-9.]+)\s+(\w+)\s+([0-9.]+)\s+(\w+)\]'
    throughput_pattern = r'thrpt:\s+\[([0-9.]+)\s+(\w+)\s+([0-9.]+)\s+(\w+)\s+([0-9.]+)\s+(\w+)\]'
    
    for match in re.finditer(bench_pattern, log_content):
        name = match.group(1)
        lower_bound = float(match.group(2))
        estimate = float(match.group(4))
        upper_bound = float(match.group(6))
        unit = match.group(5)
        
        metrics[name] = {
            'time': {
                'estimate': estimate,
                'lower_bound': lower_bound,
                'upper_bound': upper_bound,
                'unit': unit
            }
        }
    
    return metrics

def analyze_memory_patterns(log_content: str) -> Dict[str, Any]:
    """Extract memory allocation patterns from logs"""
    patterns = {}
    
    # Extract allocation statistics
    alloc_pattern = r'(\w+(?:\s+\w+)*) - Allocations: (\d+), Total bytes: (\d+), Allocs/op: ([0-9.]+), Bytes/op: ([0-9.]+)'
    
    for match in re.finditer(alloc_pattern, log_content):
        method = match.group(1)
        allocations = int(match.group(2))
        total_bytes = int(match.group(3))
        allocs_per_op = float(match.group(4))
        bytes_per_op = float(match.group(5))
        
        patterns[method] = {
            'allocations': allocations,
            'total_bytes': total_bytes,
            'allocs_per_op': allocs_per_op,
            'bytes_per_op': bytes_per_op
        }
    
    return patterns

def generate_insights(serialization_metrics: Dict, memory_patterns: Dict) -> List[str]:
    """Generate optimization insights from benchmark data"""
    insights = []
    
    # Analyze serialization performance
    if 'serialization_methods/medium_to_be_bytes' in serialization_metrics:
        vec_time = serialization_metrics['serialization_methods/medium_to_be_bytes']['time']['estimate']
        
        if 'serialization_methods/medium_to_be_bytes_buf' in serialization_metrics:
            bytes_time = serialization_metrics['serialization_methods/medium_to_be_bytes_buf']['time']['estimate']
            improvement = vec_time / bytes_time
            insights.append(f"Bytes approach is {improvement:.2f}x faster than Vec approach")
        
        if 'serialization_methods/medium_raw_pointer' in serialization_metrics:
            raw_time = serialization_metrics['serialization_methods/medium_raw_pointer']['time']['estimate']
            improvement = vec_time / raw_time
            insights.append(f"Raw pointer approach is {improvement:.2f}x faster than Vec approach")
    
    # Analyze memory efficiency
    if 'Vec approach' in memory_patterns and 'Bytes approach' in memory_patterns:
        vec_allocs = memory_patterns['Vec approach']['allocs_per_op']
        bytes_allocs = memory_patterns['Bytes approach']['allocs_per_op']
        
        if vec_allocs > bytes_allocs:
            reduction = (vec_allocs - bytes_allocs) / vec_allocs * 100
            insights.append(f"Bytes approach reduces allocations by {reduction:.1f}%")
    
    # Identify optimization opportunities
    max_allocs_per_op = max([p['allocs_per_op'] for p in memory_patterns.values()] + [0])
    if max_allocs_per_op > 2:
        insights.append("High allocation frequency detected - consider buffer reuse patterns")
    
    return insights

def main():
    """Main analysis function"""
    print("🔍 Analyzing benchmark results...")
    
    # Parse log files
    results = {}
    
    for log_file in Path('.').glob('*_output.log'):
        print(f"Processing {log_file}...")
        content = log_file.read_text()
        
        if 'serialization' in log_file.name:
            results['serialization'] = parse_criterion_output(content)
        elif 'memory_allocation' in log_file.name:
            results['memory'] = analyze_memory_patterns(content)
    
    # Generate insights
    serialization_metrics = results.get('serialization', {})
    memory_patterns = results.get('memory', {})
    insights = generate_insights(serialization_metrics, memory_patterns)
    
    # Write analysis report
    with open('analysis_report.json', 'w') as f:
        json.dump({
            'serialization_metrics': serialization_metrics,
            'memory_patterns': memory_patterns,
            'insights': insights
        }, f, indent=2)
    
    # Print summary
    print("\n📊 Analysis Summary:")
    for insight in insights:
        print(f"  • {insight}")
    
    print(f"\n✅ Detailed analysis saved to analysis_report.json")

if __name__ == '__main__':
    main()
EOF

chmod +x "${RESULTS_DIR}/analyze_results.py"

# Create comparison script for tracking performance over time
cat > "${RESULTS_DIR}/compare_with_baseline.sh" << 'EOF'
#!/bin/bash

# Compare current results with baseline performance
# Usage: ./compare_with_baseline.sh [baseline_dir]

BASELINE_DIR=${1:-"../baseline_results"}

if [ ! -d "$BASELINE_DIR" ]; then
    echo "❌ Baseline directory not found: $BASELINE_DIR"
    echo "Run benchmarks first to establish baseline"
    exit 1
fi

echo "🔄 Comparing with baseline results in $BASELINE_DIR"

# Compare serialization performance
if [ -f "serialization_output.log" ] && [ -f "$BASELINE_DIR/serialization_output.log" ]; then
    echo "📈 Serialization Performance Changes:"
    # Extract key metrics and compare
    # This would be enhanced with proper metric extraction
    echo "  (Detailed comparison would require metric extraction tool)"
fi

# Compare memory usage
if [ -f "memory_allocation_output.log" ] && [ -f "$BASELINE_DIR/memory_allocation_output.log" ]; then
    echo "💾 Memory Usage Changes:"
    echo "  (Detailed comparison would require metric extraction tool)"
fi

echo "✅ Comparison complete"
EOF

chmod +x "${RESULTS_DIR}/compare_with_baseline.sh"

# Run analysis if Python is available
if command -v python3 &> /dev/null; then
    echo -e "${YELLOW}Running automated analysis...${NC}"
    cd "${RESULTS_DIR}"
    python3 analyze_results.py
    cd ..
else
    echo -e "${YELLOW}Python3 not found - skipping automated analysis${NC}"
fi

echo -e "${GREEN}🎉 Benchmark suite completed!${NC}"
echo -e "${BLUE}Results saved in: ${RESULTS_DIR}${NC}"
echo ""
echo -e "${YELLOW}Next steps:${NC}"
echo "1. Review HTML reports in target/criterion/"
echo "2. Check analysis_report.json for key insights"
echo "3. Use profile.sh for detailed hardware profiling"
echo "4. Compare with baseline using compare_with_baseline.sh"
echo ""
echo -e "${GREEN}For detailed profiling: ./profile.sh${NC}"