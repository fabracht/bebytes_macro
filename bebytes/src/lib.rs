#![cfg_attr(not(feature = "std"), no_std)]

#[cfg(feature = "std")]
extern crate std;

#[cfg(not(feature = "std"))]
extern crate alloc;
#[cfg(not(feature = "std"))]
pub use alloc::borrow::ToOwned;

// Re-export Vec for use in generated code
#[cfg(not(feature = "std"))]
pub use alloc::vec::Vec;
#[cfg(feature = "std")]
pub use std::vec::Vec;

pub mod buffer;
pub mod interpreter;

pub use bebytes_derive::BeBytes;
#[doc(hidden)]
pub use interpreter::{StringInterpreter, Utf8};

// Re-export buffer types for generated code (previously from bytes crate)
pub use buffer::{BufMut, Bytes, BytesMut};

/// Error type for `BeBytes` operations
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BeBytesError {
    /// Buffer is empty when data was expected
    EmptyBuffer,
    /// Not enough data in buffer
    InsufficientData { expected: usize, actual: usize },
    /// Invalid enum discriminant value
    InvalidDiscriminant { value: u8, type_name: &'static str },
    /// Bit field value exceeds maximum allowed
    InvalidBitField {
        value: u128,
        max: u128,
        field: &'static str,
    },
}

impl core::fmt::Display for BeBytesError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::EmptyBuffer => write!(f, "No bytes provided"),
            Self::InsufficientData { expected, actual } => {
                write!(f, "Not enough bytes: expected {expected}, got {actual}")
            }
            Self::InvalidDiscriminant { value, type_name } => {
                write!(f, "Invalid discriminant {value} for type {type_name}")
            }
            Self::InvalidBitField { value, max, field } => {
                write!(f, "Value {value} exceeds maximum {max} for field {field}")
            }
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for BeBytesError {}

// Note: core::error::Error is stable since Rust 1.81
// We could add this when we update MSRV:
// #[cfg(not(feature = "std"))]
// impl core::error::Error for BeBytesError {}

pub trait BeBytes {
    fn field_size() -> usize;

    // Big-endian methods
    #[cfg(feature = "std")]
    fn to_be_bytes(&self) -> std::vec::Vec<u8>;

    #[cfg(not(feature = "std"))]
    fn to_be_bytes(&self) -> alloc::vec::Vec<u8>;

    /// Try to parse a struct from big-endian bytes
    ///
    /// # Errors
    ///
    /// Returns `BeBytesError::EmptyBuffer` if the input slice is empty
    /// Returns `BeBytesError::InsufficientData` if there aren't enough bytes to parse all fields
    /// Returns `BeBytesError::InvalidDiscriminant` if an enum field has an invalid value
    /// Returns `BeBytesError::InvalidBitField` if a bit field value exceeds its maximum
    fn try_from_be_bytes(bytes: &'_ [u8]) -> core::result::Result<(Self, usize), BeBytesError>
    where
        Self: Sized;

    // Little-endian methods
    #[cfg(feature = "std")]
    fn to_le_bytes(&self) -> std::vec::Vec<u8>;

    #[cfg(not(feature = "std"))]
    fn to_le_bytes(&self) -> alloc::vec::Vec<u8>;

    /// Convert to big-endian bytes as a Bytes buffer
    ///
    /// Returns a Bytes buffer (now a simple Vec wrapper without reference counting)
    fn to_be_bytes_buf(&self) -> Bytes {
        // Default implementation for backward compatibility
        Bytes::from(self.to_be_bytes())
    }

    /// Convert to little-endian bytes as a Bytes buffer
    ///
    /// Returns a Bytes buffer (now a simple Vec wrapper without reference counting)
    fn to_le_bytes_buf(&self) -> Bytes {
        // Default implementation for backward compatibility
        Bytes::from(self.to_le_bytes())
    }

    /// Try to parse a struct from little-endian bytes
    ///
    /// # Errors
    ///
    /// Returns `BeBytesError::EmptyBuffer` if the input slice is empty
    /// Returns `BeBytesError::InsufficientData` if there aren't enough bytes to parse all fields
    /// Returns `BeBytesError::InvalidDiscriminant` if an enum field has an invalid value
    /// Returns `BeBytesError::InvalidBitField` if a bit field value exceeds its maximum
    fn try_from_le_bytes(bytes: &'_ [u8]) -> core::result::Result<(Self, usize), BeBytesError>
    where
        Self: Sized;

    /// Encode directly to a buffer in big-endian format
    ///
    /// Writes struct data directly to the provided buffer without intermediate
    /// allocations.
    ///
    /// # Errors
    ///
    /// Returns an error if the buffer doesn't have enough capacity
    fn encode_be_to<B: BufMut>(&self, buf: &mut B) -> core::result::Result<(), BeBytesError> {
        // Default implementation using to_be_bytes for backward compatibility
        let bytes = self.to_be_bytes();
        buf.put_slice(&bytes);
        Ok(())
    }

    /// Encode directly to a buffer in little-endian format
    ///
    /// Writes struct data directly to the provided buffer without intermediate
    /// allocations.
    ///
    /// # Errors
    ///
    /// Returns an error if the buffer doesn't have enough capacity
    fn encode_le_to<B: BufMut>(&self, buf: &mut B) -> core::result::Result<(), BeBytesError> {
        // Default implementation using to_le_bytes for backward compatibility
        let bytes = self.to_le_bytes();
        buf.put_slice(&bytes);
        Ok(())
    }
}
