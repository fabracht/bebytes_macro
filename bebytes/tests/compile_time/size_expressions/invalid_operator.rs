// This test verifies that invalid operators in size expressions
// produce a compile-time error.

use bebytes::BeBytes;
#[cfg(not(feature = "std"))]
extern crate alloc;

#[derive(BeBytes, Debug, PartialEq)]
struct InvalidOperator {
    length: u16,
    // Error: Bitwise operators not supported in size expressions
    #[bebytes(size = "length & 0xFF")]
    data: Vec<u8>,
}

fn main() {}