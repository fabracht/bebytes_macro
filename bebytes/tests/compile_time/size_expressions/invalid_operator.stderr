error: Invalid size expression: Unsupported binary operator
  --> tests/compile_time/size_expressions/invalid_operator.rs:12:5
   |
12 |     #[bebytes(size = "length & 0xFF")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
